### PR TITLE
Vocalize word formatting shortcuts

### LINKED ISSUE #10271: precise Announcing of the state of formatting settings when a hotkey in word is pressed
### Is your feature request related to a problem? Please describe.
No, not in the sense of a bug. It's just something I miss.
### Describe the solution you'd like
In word, for example, when I press the keystroke for bold, I would like to have NVDA say "bold on". Or "bold off", depending on what it is set to currently.
### Describe alternatives you've considered
Of course, I could just do it and check formating using NVDA+f, but it's just easier this way. Also I have a more direct confirmation.

### PR DESCRIPTION
**Pull request by CyrilleB79**
### Link to issue number:

Fixes #10271 

### Summary of the issue:

User wants to have font formatting state announced when native font formatting shortcut are executed under Word and Outlook.

#### General description

This PR allows to announce the new state of font formatting when the following shortcuts are executed in MS Word (English version):

* `control+shift+D`: Underline double
* `control+shift+W`: Underline words only
* `control+shift+A`: All caps
* Ctrl+Shift+K: Small caps

Moreover it announce the result of the `shift+F3` shortcut that change character case.

At last it include `control+plus` and `control+shift+plus` that are alternative shortcuts for subscript and superscript.

#### Notes about various capitalization shortcuts

* In Word, the Shift+F3 and Ctrl+Shift+A shortcuts are slightly different:

  * Shift+F3 only modifies the case of the current selection (or the word under the cursor if there is no selection)
  * Ctrl+Shift+A also modifies the case of selected text (or current word), but in addition, it modifies the case of the characters that will be typed at insertion point of the cursor.

* Outlook's Ctrl+Shift+A shortcut is equivalent to Word's Shift+F3, whereas Word's Ctrl+Shift+A has no equivalent under Outlook as far as I know.

#### Messages

Word's shortcuts Ctrl+Shift+A, Ctrl+Shift+K and Shift+F3 modify all the selection's (or current word's) capitalization. However, Ctrl+Shift+A and Ctrl+Shift+K also impact the next characters that will be typed. Thus for this shortcut "on" and "off" have been used in their announced message in order to denote a state change. On the contrary Shift+F3 announces only the result of a modification of current selection (or word); thus "on" and "off" have not been used in its message.

### Testing performed:

Tested all combinations:

* In Word and in Outlook message (edition), English version
* For all previously listed shortcuts
* With the following selections:

  * No selection, cursor in the middle of a word
  * No selection, cursor among case insensitive characters (numbers)
  * One single letter selected
  * Two letters selected
  * One word selected
  * Many words selected, without punctuation in middle of them
  * Sentence (many words followed by period) selected

Regarding Shift+F3, I am using a Microsoft enumeration containing also values such as wdHalfWidth, wdFullWidth, wdKatakana and wdHiragana. Seems all concepts of far-east languages.
Is there a notion equivalent to case changing in these languages? Do Shift+F3 or other shortcuts unknown to me have an impact on far-east writing?
Maybe @josephsl and @nishimotz could comment and test.

### Known issues with pull request:

* Translators will have to modify the `gestures.ini` files in SRT repo (or elsewhere if the process changes) according to native Word or Outlook shortcuts in their language if these differ from English. I plan to inform them in detail when the beta translation freeze occurs.
* `ctrl+shift+I`, `ctrl+shift+b` and `ctrl+shift+U` are alternative shortcuts for italic, bold and underline. However, since they trigger UIA notifications. So, as per https://github.com/nvaccess/nvda/pull/10283#issuecomment-883861903, I do not include them in this PR. This can be revisited when #10990  is merged.

## Summary by CodeRabbit

- **New Features**
  - Enhanced vocalization of font formatting shortcuts in Word and Outlook, including underline styles and character case types.
  
- **Bug Fixes**
  - Added specific keyboard shortcuts handling for Outlook and Word to improve user interaction.

### CODE REVIEW

**Comment by bdorer** [comment]
is there any progress here?

**Comment by CyrilleB79** [comment]
On my side, the PR is ready.
Feel free to test it (maybe the gesture.ini file must be updated according to your locale) and provide feedback if you want.
This PR has been labelled as feature by @feerrenrut. Maybe such PRs are not prioritarised these days by NVAccess. Let's see what is happening when #11006 is resolved.

**Comment by CyrilleB79** [comment]
I am adding two additional shortcut keys for toggleSuperscriptSubscript script:
* kb:control+plus
* kb:control+shift+plus
Indeed, they are also valid shortcuts key to format the text to superscript or subscript in English version of Word or Outlook, as well as in some other versions (e.g. French, Italian). These shortcuts cannot be executed with English (US) keyboard, but can be executed in English Word with another keyboard allowing to type "+" without any modifier, such as German or Italian.

I will update the initial description of this PR accordingly.

File: source/NVDAObjects/window/winword.py
Comment (feerrenrut) [inline source/NVDAObjects/window/winword.py:78]: It is obvious to some, and it was to me in the past, but right now I can't remember what the best way is to find these constants. Could you add a comment at the top of the constants list describing how best to find them.

File: source/NVDAObjects/window/winword.py
Comment (feerrenrut) [inline source/NVDAObjects/window/winword.py:1449]: It may be that the first operation hasn't yet completed. It would probably be best if nvda scripts queued subsequent runs until the first is complete. Could you see if you can confirm if that is happening here?

**Review by feerrenrut — CHANGES_REQUESTED** [review]
What about if the users customizes keyboard shortcuts in word?

**Comment by CyrilleB79** [comment]
@feerrenrut wrote:
> What about if the users customizes keyboard shortcuts in word?

This is not supported in this PR. This PR uses more or less the same strategy than what already exists in NVDA to report text formatting shortcuts in Word such as italic, underline and bold. Thus it suffers from the same limitations. If the user changes the default shortcuts in Word, the result of the new shortcut will not be reported.
I do not know if many users change default shortcuts.
If the user is a dev / tech-friendly people, he may modify manually his gestures.ini file, as translators do. But in any case, it is not an operation that any user can execute.

File: source/NVDAObjects/window/winword.py
Comment (CyrilleB79) [reply source/NVDAObjects/window/winword.py:1451]: How do you perform this? I.e. how do you delay the second script until the end of the first?

Moreover, I have tried to reproduce the issue with only one polling, but I cannot anymore. I will try on another version of Word later.

Comment (CyrilleB79) [reply source/NVDAObjects/window/winword.py:1451]: I have just tested on another Word and Outlook version and I cannot reproduce the issue either. Both Office are 2016, but one of them was not up-to-date.

**Comment by feerrenrut** [comment]
> This is not supported in this PR. This PR uses more or less the same strategy than what already exists in NVDA to report text formatting shortcuts in Word such as italic, underline and bold. Thus it suffers from the same limitations. 

Ok, fair point this is not a new concern.

File: source/NVDAObjects/window/winword.py
Comment (feerrenrut) [reply source/NVDAObjects/window/winword.py:1451]: >How do you perform this? I.e. how do you delay the second script until the end of the first?

We'd need to first confirm that is what is happening, my suggestion was really just a guess. Since I didn't know, I have done some investigation, the short version is that it seems unlikely that the script is being called again before it completes. The long answer follows:

To work it out we can look at how the gestures are called. 

It all starts with the `hookThreadFunc` in `source/winInputHook.py`. This runs in it's own thread, it registers a Low level hook function for the keyboard. Unfortunately this is where the questions start: 
- Is this callback `keyboardHook` called only in the context of the thread that installed it? 
  - The [docs for "LowLevelKeyboardProc callback function"](https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/ms644985(v=vs.85)) say:
  > This hook is called in the context of the thread that installed it. The call is made by sending a message to the thread that installed the hook. Therefore, the thread that installed the hook must have a message loop.
- But the ["Hooks Overview" docs ](https://docs.microsoft.com/en-us/windows/win32/winmsg/about-hooks) say the following, please note that [`WH_KEYBOARD_LL` is a "global only" install](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowshookexw):
 >  A global hook procedure can be called in the context of any application in the same desktop as the calling thread, so the procedure must be in a separate DLL module. 
- Can this callback be re-entered by the Windows system?
  - I couldn't find any information about this. 

Another interesting piece of advice from the [docs for "LowLevelKeyboardProc callback function"](https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/ms644985(v=vs.85)) which we are apparently ignoring is:
>Note  Debug hooks cannot track this type of low level keyboard hooks. If the application must use low level hooks, it should run the hooks on a dedicated thread that passes the work off to a worker thread and then immediately returns. In most cases where the application needs to use low level hooks, it should monitor raw input instead. This is because raw input can asynchronously monitor mouse and keyboard messages that are targeted for other threads more effectively than low level hooks can. For more information on raw input, see Raw Input.

Ok, well I say "ignoring", but it looks like most paths result in adding an event to the event queue. The logic is complicated enough that it is hard to confirm this happens for all paths. If we assume it does happen for all paths, then I think we can be fairly confident that the script can't be called until it is exited.

I'm going to call on @michaelDCurran to take a look at this to see if there is anything I might have missed here, or if he can clarify anything further.

Comment (CyrilleB79) [reply source/NVDAObjects/window/winword.py]: @feerrenrut thanks for your detailed explanations (and also for the summary!).

I have tried again but I cannot reproduce the issue anymore, This PR is quite old (it was one of my first PRs) and I do not remember exactly which type of stress test may have produced the issue; I also do not remember for example if I had tried to restart my computer.
I have thus removed this double polling. If an issue occur we my investigate on this new basis.

Comment (CyrilleB79) [reply source/NVDAObjects/window/winword.py]: @feerrenrut, have you had the opportunity to discuss with @michaelDCurran on the double script call potential issue.
Since I have removed the double-polling anyway, this conversation is not blocked anymore IMO. Do you agree?

**Comment by lukaszgo1** [comment]
@feerrenrut  @michaelDCurran  Would it be possible to move this PR forward?

**Review by feerrenrut — CHANGES_REQUESTED** [review]
Overall the code looks good.

But testing this locally looking speech viewer:
- Pressing `ctrl+u` shows two lines : 
   ```
   Underline Single
   Underline On
   ```
- Pressing `ctrl+b` shows two lines, note the difference in case:
   ```
   Bold on
   Bold On
   ```
Only the second sequence is heard, meaning that some information is lost in the underline case.
I haven't tested the other shortcuts yet, but I think this should be addressed.

**Comment by josephsl** [comment]
Hi, the seemingly duplicate lines are caused by UIA notification event (see #10950 for details). Thanks.

**Comment by CyrilleB79** [comment]
Regarding duplicate format reporting, the issue is not specific to this PR, as explained by @josephsl.

Looking again at this old PR, I am not so happy with the announcement "Underline single" instead of today's "Underline on". Maybe in the case of much more common single underline we can just keep "Underline on".
What do you think?

**Comment by feerrenrut** [comment]
> Regarding duplicate format reporting, the issue is not specific to this PR,

Even though it is not specific to this PR, it is a conflict in the UI. The new messages may not be heard, it is also causing NVDA to do more than it needs to.

>Looking again at this old PR, I am not so happy with the announcement "Underline single" instead of today's "Underline on". Maybe in the case of much more common single underline we can just keep "Underline on".
What do you think?

I'm not sure, I assume that the other values might be "underline double", any others? This provides more information than just "underline on".

**Comment by CyrilleB79** [comment]
> > Regarding duplicate format reporting, the issue is not specific to this PR,
> 
> Even though it is not specific to this PR, it is a conflict in the UI. The new messages may not be heard, it is also causing NVDA to do more than it needs to.

I may have not been clear enough in my previous comment. Let me give some more details:
1. I have understood that there is already a PR from @josephsl to fix the double reporting issue: #10990. @josephsl can you confirm this? If yes, there is no point in fixing the issue in this PR.
2. In my version of office, the UIA notifications are raised only only upon the following key press and only in MS Word (not Outlook): ctrl+I (italic), ctrl+U (underline) and ctrl+G (French equivalent of ctrl+B = bold). There is no UIA notification for the various capitalization shortcuts, the other underline shortcuts (ctrl+shift+D = "double" and ctrl+shift+W = "words only"). There is also no UIA notification for the following alternative shortcuts: ctrl+shift+I/U/B, ctrl+plus and ctrl+shift+plus.
To summarize, none of the new shortcut key introduced in this PR have the double reporting issue in my version of Office.
@joseph, maybe however the set of shortcuts that report UIA notification has been extended in newer versions of MS Office. Could you report the result withe all the shortcuts listed in this PR's description? Thanks.

So IMO, this PR can be progressed independantly from #10990.
Anyway, if you have another opinion, I would prefer to attribute a blocked label to this PR, until #10990 is merged.

**Comment by CyrilleB79** [comment]
> > Looking again at this old PR, I am not so happy with the announcement "Underline single" instead of today's "Underline on". Maybe in the case of much more common single underline we can just keep "Underline on".
> > What do you think?
> 
> I'm not sure, I assume that the other values might be "underline double", any others? This provides more information than just "underline on".

OK I let you the final decision on this point. So unless there is another comment regarding this topic, I keep the announcement "Underline single".
The other underline styles that have a Word shortcut key are: underline double (control+shif+D) and underline words only (control+shift+W). But I guess they are much less commonly used.

**Comment by josephsl** [comment]
Hi, the PR mentioned actually does the opposite: it will silence these UIA notifications altogether. This is in consideration for those running Office in older Windows releases (7 and 8.1 to be exact). Ideally UIA notification is the way to go, but given not all users are on latest releases, a compromise was proposed in the mentioned PR. Thanks.

**Comment by CyrilleB79** [comment]
> Hi, the PR mentioned actually does the opposite: it will silence these UIA notifications altogether. This is in consideration for those running Office in older Windows releases (7 and 8.1 to be exact). Ideally UIA notification is the way to go, but given not all users are on latest releases, a compromise was proposed in the mentioned PR. Thanks.

If I understand correctly, #10990 silences UIA notification regarding font modification on whatever version of Office or Windows. So this should be enough to suppress double font modification reporting. Can you confirm this @josephsl?

Of course, I understand that in the future, using UIA notification should be the recommended way to do, at least when these notifications have reached a sufficent level of maturity. In Office 2016 this is clearly not yet the case since:
* only a subset of font formatting shortcuts raise UIA notifications (ctrl+I/U/G=B)
* the notifications are raised only in Word, not in Outlook
* I guess the notifications are not localized in French (unless it has changed recently, I have to double-check)

When we implement font formatting reporting, I think we should determine the source of the reporting (object model or UIA) according to Windows and Office version.
But this should be done in another PR and the current one should concentrate ont the object model part.

**Comment by josephsl** [comment]
Hi, what Cyrille raises is the exact comment @MichaelDCurran talks about in more recent comments for #10990. As for suppressing UIA notifications, Office will raise this event if one is running recent Word releases under Windows 10 1709 (Fall Creators Update) and later (UIA notification event was introduced in 1709). Thanks.

**Comment by feerrenrut** [comment]
It seems like there are some complexities here. Someone will need to deep dive on this to
- [ ] enumerate the different situations, whether UIA and object model are available.
- [ ] Are UIA notifications translated?

In situations where both object model and UIA notifications are available, a setting to select between them would be a safe fallback. In other situations the appModule should select what type of notification system is used. This may need to be at individual event granularity. 

Given the clash here, I don't think this can be merged until we have these answers. @CyrilleB79 I think we should convert this PR back to draft. Perhaps you and @josephsl can work together on reaching a solution for this?

**Comment by CyrilleB79** [comment]
@feerrenrut I do not understand if you disagree with or if you just missed my proposal to handle the two points in separate PRs:
* Support more font formatting shortcut reporting via object model (as done today); this is the object of the current PR
* Deal with the issue of potential double reporting, i.e. choose when to use object model and when to use UIA: should be implemented in a new PR

IMO it's more clear that these two tasks be handled separately, i.e. in 2 PRs.

And I think they do not depend so much on each other. Anyway if you disagree on this last point, we may set this PR on hold while the other one (UIA vs object model) is implemented and merged. Just let me know before I continue on these topics.

**Comment by feerrenrut** [comment]
When we are in doubt about a PR we opt towards no change. There are several open questions here, the resulting user experience of this change can't be determined until they are answered. I'm concerned that it isn't certain which message a user will receive (UIA or object model).

I would prefer to hold off on merging this PR until these questions are answered or resolved. I'm happy for that to happen via another PR. Until then I'll mark this blocked.

**Comment by CyrilleB79** [comment]
OK. Thanks for this clarification.
I will try to have a look on the other topic with Joseph (if he can) or someone else since I do not have a Word 365 or 2019 to test with the most recent versions of Office.

> I'm concerned that it isn't certain which message a user will receive (UIA or object model).
 
Regarding this point, the situation is the following in the current version of NVDA (unchanged with this PR):
* In the case where the UIA notifications cannot be reported (i.e. Windows version < Windows 10 1709 or older version of Word), only NVDA's notification based on the object model is reported
* In the case UIA can be reported, the notification based on the object model begins to be reported, but is interrupted by the UIA notification. So actually, in this case, only the UIA notification can be fully heard, but a little part of the object model notification can sometimes be heard beforehand.

**Comment by seanbudd** [comment]
Closing as abandoned

**Comment by CyrilleB79** [comment]
@seanbudd, this PR is not abandoned but blocked by #10990. I agree that this is quite old.

In the meantime we have been able to see the evolution of UIA notifications regarding text formatting. After all these years, Microsoft has not added new notifications, i.e. only bold, italic and underline are being reported by UIA notifications.

So to progress this PR, I propose to remove the changes related to bold, italic and underline and underline, i.e. all the shortcuts conflicting with UIA notifications, since it's the goal of #10990.

@seanbudd do you agree with this plan? If not, why?

Reopening and removing the "Abandoned" label. We may close again after discussion if needed.

**Comment by seanbudd** [comment]
this plan makes sense, my assumption was that the work was abandoned. thanks

**Comment by seanbudd** [comment]
I'm marking this as ready for review with the note that it is currently blocked

**Comment by CyrilleB79** [comment]
@seanbudd with my last changes, I do not consider this PR blocked anymore by #10990.

However, I still have to re-test all the use cases; it seems to me that not all of them are functional. That's why I kept this PR as draft. I will also address the Rabbit's review and let you know when it is ready.

Or was there any other reason why you consider this PR blocked?

**Comment by seanbudd** [comment]
Thanks, good to know, it was a misunderstanding on my part

**Comment by seanbudd** [comment]
@CyrilleB79 - do you plan to do any further work here?

**Comment by CyrilleB79** [comment]
Due to file reformatting, I am not able to upmerge from master.

Finally, I have opened #17197 with a new branch as a replacement of this PR.

So I am closing in favor of #17197.

**Pull request closed without merging**
